src/Silex/AppArgumentValueResolver.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+/*
+ * This file is part of the Silex framework.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Silex;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Controller\ArgumentValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+
+/**
+ * HttpKernel Argument Resolver for Silex.
+ *
+ * @author Romain Neutron <imprec@gmail.com>
+ */
+class AppArgumentValueResolver implements ArgumentValueResolverInterface
+{
+    private $app;
+
+    public function __construct(Application $app)
+    {
+        $this->app = $app;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request, ArgumentMetadata $argument)
+    {
+        return $argument->getType() === Application::class;
```

Comment:
That will return `false` if the type hint is made on a subclass of `Application`.

```php
use Silex\Application;
class MyApp extends Application {}
$app = new MyApp;
$app->get('/', function (MyApp $app) { return 'Hello'; });
$app>->run();
```

This might not be considered as a supported feature, but it is certainly a breaking change.